[Watch the YouTube Video](https://www.youtube.com/watch?v=8z5jsznoB0A) <p>

![Description of the graphic](graphic.jpg)


In [30]:
import pandas as pd
import plotly.express as px

# Step 1: Load the data
df_2022 = pd.read_csv('Police_Incidents_2022.csv')
df_2023 = pd.read_csv('Police_Incidents_2023.csv')



In [31]:
# Step 2: Data cleaning (this is a basic check)
print(df_2022.info())
print(df_2023.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26526 entries, 0 to 26525
Data columns (total 23 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   publicaddress      26526 non-null  object 
 1   caseNumber         26526 non-null  object 
 2   precinct           26526 non-null  object 
 3   reportedDate       26526 non-null  object 
 4   reportedTime       26526 non-null  int64  
 5   beginDate          26526 non-null  object 
 6   reportedDateTime   26526 non-null  object 
 7   beginTime          26526 non-null  int64  
 8   offense            26526 non-null  object 
 9   description        26526 non-null  object 
 10  UCRCode            25992 non-null  float64
 11  enteredDate        26526 non-null  object 
 12  centergbsid        0 non-null      float64
 13  centerLong         26526 non-null  float64
 14  centerLat          26526 non-null  float64
 15  centerX            26526 non-null  float64
 16  centerY            265

In [32]:
# Distinct list of offenses in 2022 data
offenses_2022 = df_2022['offense'].unique()
print("Distinct offenses in 2022:")
print(offenses_2022)

# Distinct list of offenses in 2023 data
offenses_2023 = df_2023['offense'].unique()
print("\nDistinct offenses in 2023:")
print(offenses_2023)


Distinct offenses in 2022:
['THEFT' 'TMVP' 'TFMV' 'BIKETF' 'SHOPLF' 'BURGD' 'THFTSW' 'BURGB' 'ROBPER'
 'AUTOTH' 'TBLDG' 'ONLTHT' 'ROBPAG' 'WLKOFF' 'ROBBIZ' 'ASLT2' 'CSCR'
 'DASLT2' 'CARJCK' 'DASLT3' 'ASLT3' 'DASTR' 'TFPER' 'MURDR' 'ASLT4'
 'ARSON' 'ASLT1' 'ADLTTN' 'MVTHFT' 'DISARM' 'COMPUT' 'JHOMIC' 'PETIT'
 'NOPAY' 'SCRAP' 'COINOP' 'DASLT1' 'POCKET' 'ABSVAR']

Distinct offenses in 2023:
['CSCR' 'CARJCK' 'AUTOTH' 'THEFT' 'BURGD' 'TFMV' 'TMVP' 'ARSON' 'BURGB'
 'SHOPLF' 'DASLT3' 'ROBPAG' 'ASLT2' 'DASTR' 'ROBPER' 'ASLT4' 'TBLDG'
 'THFTSW' 'TFPER' 'DASLT2' 'BIKETF' 'ASLT3' 'ROBBIZ' 'ADLTTN' 'MURDR'
 'WLKOFF' 'ONLTHT' 'PETIT' 'MVTHFT' 'COINOP' 'ASLT1' 'COMPUT' 'SCRAP'
 'POCKET' 'DISARM' 'DASLT1' 'NOPAY' 'JHOMIC' 'ABSVAR']


In [33]:
# Updated violent and non-violent classification based on formal law enforcement categories
violent_crimes = ['ASLT1', 'ASLT2', 'ASLT3', 'ASLT4', 'DASLT1', 'DASLT2', 'DASLT3', 'DASTR', 'CARJCK', 
                  'ROBPER', 'ROBBIZ', 'ROBPAG', 'MURDR', 'JHOMIC', 'CSCR', 'ADLTTN']
non_violent_crimes = ['THEFT', 'TFMV', 'TMVP', 'BIKETF', 'SHOPLF', 'THFTSW', 'ONLTHT', 'TBLDG', 'AUTOTH', 
                      'TFPER', 'MVTHFT', 'WLKOFF', 'BURGD', 'BURGB', 'ARSON', 'COMPUT', 'SCRAP', 'COINOP', 
                      'POCKET', 'NOPAY', 'DISARM', 'PETIT']

# Function to classify based on new categories
def categorize_crime(offense):
    if offense in violent_crimes:
        return 'Violent'
    elif offense in non_violent_crimes:
        return 'Non-Violent'
    else:
        return 'Other'

# Apply classification to both datasets
df_2022['Crime_Category'] = df_2022['offense'].apply(categorize_crime)
df_2023['Crime_Category'] = df_2023['offense'].apply(categorize_crime)

# Check classification counts
print("2022 Crime Category Counts:")
print(df_2022['Crime_Category'].value_counts())

print("2023 Crime Category Counts:")
print(df_2023['Crime_Category'].value_counts())


2022 Crime Category Counts:
Crime_Category
Non-Violent    21914
Violent         4610
Other              2
Name: count, dtype: int64
2023 Crime Category Counts:
Crime_Category
Non-Violent    21768
Violent         4244
Other              1
Name: count, dtype: int64


In [34]:
# Updated violent crimes list to include ABSVAR
violent_crimes = ['ASLT1', 'ASLT2', 'ASLT3', 'ASLT4', 'DASLT1', 'DASLT2', 'DASLT3', 'DASTR', 'CARJCK', 
                  'ROBPER', 'ROBBIZ', 'ROBPAG', 'MURDR', 'JHOMIC', 'CSCR', 'ADLTTN', 'ABSVAR']

non_violent_crimes = ['THEFT', 'TFMV', 'TMVP', 'BIKETF', 'SHOPLF', 'THFTSW', 'ONLTHT', 'TBLDG', 'AUTOTH', 
                      'TFPER', 'MVTHFT', 'WLKOFF', 'BURGD', 'BURGB', 'ARSON', 'COMPUT', 'SCRAP', 'COINOP', 
                      'POCKET', 'NOPAY', 'DISARM', 'PETIT']

# Function to classify based on updated categories
def categorize_crime(offense):
    if offense in violent_crimes:
        return 'Violent'
    elif offense in non_violent_crimes:
        return 'Non-Violent'
    else:
        return 'Other'

# Apply the updated classification function again
df_2022['Crime_Category'] = df_2022['offense'].apply(categorize_crime)
df_2023['Crime_Category'] = df_2023['offense'].apply(categorize_crime)

# Check the updated counts to verify the change
print("2022 Crime Category Counts (Updated):")
print(df_2022['Crime_Category'].value_counts())

print("2023 Crime Category Counts (Updated):")
print(df_2023['Crime_Category'].value_counts())


2022 Crime Category Counts (Updated):
Crime_Category
Non-Violent    21914
Violent         4612
Name: count, dtype: int64
2023 Crime Category Counts (Updated):
Crime_Category
Non-Violent    21768
Violent         4245
Name: count, dtype: int64


```python
# I made this markdown, but used it to determine all "other" categories from above and correctly characterize them.

# Investigate the 'Other' category in 2022
other_2022 = df_2022[df_2022['Crime_Category'] == 'Other']
print("Offenses classified as 'Other' in 2022:")
print(other_2022[['caseNumber', 'offense', 'description']])

# Investigate the 'Other' category in 2023
other_2023 = df_2023[df_2023['Crime_Category'] == 'Other']
print("\nOffenses classified as 'Other' in 2023:")
print(other_2023[['caseNumber', 'offense', 'description']])


In [35]:
import plotly.express as px

# Data preparation for visualization
crime_data = {
    'Year': ['2022', '2022', '2023', '2023'],
    'Crime_Type': ['Violent', 'Non-Violent', 'Violent', 'Non-Violent'],
    'Count': [4610, 21914, 4244, 21768]
}
df_crime_data = pd.DataFrame(crime_data)

# Bar chart comparison
fig = px.bar(df_crime_data, x='Year', y='Count', color='Crime_Type', barmode='group',
             title='Comparison of Violent and Non-Violent Crimes in 2022 vs 2023')
fig.show()


## Crime Analysis: 2022 vs. 2023

### Crime Totals:
- **2022 Total Crimes**:
  - Non-Violent: 21,914
  - Violent: 4,612
  - **Total Crimes in 2022**: 26,526

- **2023 Total Crimes**:
  - Non-Violent: 21,768
  - Violent: 4,245
  - **Total Crimes in 2023**: 26,013

---

## Calculating the Percentage Decrease in Total Crime:

To calculate the percentage decrease in total crime from 2022 to 2023, we use the formula:

$$
\text{Percentage Decrease} = \left( \frac{\text{Total in 2022} - \text{Total in 2023}}{\text{Total in 2022}} \right) \times 100
$$

Substituting the values:

$$
\text{Percentage Decrease} = \left( \frac{26526 - 26013}{26526} \right) \times 100 = \left( \frac{513}{26526} \right) \times 100 = 1.93\%
$$

Thus, the actual decrease in total crime is **1.93%**, which is significantly less than the 43% decrease claimed by the mayor.

---

### Possible Explanations for the Discrepancy:

1. **Different Crime Categories**: The mayor might have referred to a specific subset of crimes, such as violent crimes, specific types of non-violent crimes, or crimes in a particular neighborhood.

2. **Time Period**: The mayor could be comparing a specific time period (e.g., certain months or quarters) rather than the entire year.

3. **Population Adjustments**: The 43% figure could be based on crime rates adjusted for population changes. If the population has decreased significantly, it could give the appearance of a larger crime reduction.

4. **Data Gaps or Reporting Changes**: There may have been changes in how crimes were reported between 2022 and 2023. Underreporting in 2023 or increased reporting in 2022 could impact the comparison.

---

### Next Steps for Verification:

1. **Subset Analysis**: Identify if the mayor referred to a specific category of crimes (e.g., violent crimes, property crimes, or crimes in a particular area) and analyze that subset for changes.

2. **Time-Specific Analysis**: Narrow the data to specific time frames (e.g., the first 6 months of 2022 vs. 2023) to align with the period referenced by the mayor.

3. **Crime Rate Adjustments**: If population data is available, calculate the crime rate per 1,000 or 100,000 people. This could provide better insight into crime trends relative to population changes.


### Question:
Did the population of Minneapolis decrease between 2022 and 2023?

### Answer:
No, the population of Minneapolis **did not decrease** between 2022 and 2023. In fact, it saw a slight increase of around **0.74%**, growing from **422,003 in 2022** to **425,115 in 2023**. This follows a population decline from 2021 to 2022, when the city experienced a **1.36% decrease** in population.

Given this slight increase in population, changes in population are unlikely to explain the discrepancy in the mayor's claim of a **43% decrease in crime**. It's possible that the claim refers to a specific subset of crimes or time period.

Sources:
- :contentReference[oaicite:0]{index=0}ps:&#8203;:contentReference[oaicite:1]{index=1}ear/)
- [Minneapolis Population Data - World Population Review](https://worldpopulationreview.com/us-cities/minnesota/minneapolis)


### Next Steps: Exploring Crime Subcategories

In the next phase of analysis, I will:

1. **Identify the Top 4 Crime Categories** from the 2022 dataset.
2. **Compare these Top 4 Categories** with the 2023 dataset to observe trends.
3. **Categorize all other crimes** in both years as "Other."

This will provide a clearer view of which specific crime subcategories contributed the most to the overall crime rates and how these trends evolved between 2022 and 2023.


In [36]:
import pandas as pd
import plotly.express as px

# Group and count offenses for Violent and Non-Violent crimes
violent_top = df_2022[df_2022['Crime_Category'] == 'Violent']['offense'].value_counts().nlargest(4)
non_violent_top = df_2022[df_2022['Crime_Category'] == 'Non-Violent']['offense'].value_counts().nlargest(4)

# Create 'Other' category for violent crimes
violent_counts_2022 = df_2022[df_2022['Crime_Category'] == 'Violent']['offense'].value_counts()
violent_counts_2023 = df_2023[df_2023['Crime_Category'] == 'Violent']['offense'].value_counts()

violent_top_2022 = violent_counts_2022.nlargest(4)
violent_top_2023 = violent_counts_2023.nlargest(4)

violent_other_2022 = violent_counts_2022.sum() - violent_top_2022.sum()
violent_other_2023 = violent_counts_2023.sum() - violent_top_2023.sum()

# Prepare Data for Violent Crimes plot
violent_data = {
    'Year': ['2022', '2022', '2022', '2022', '2022', '2023', '2023', '2023', '2023', '2023'],
    'Offense': list(violent_top_2022.index) + ['Other'] + list(violent_top_2023.index) + ['Other'],
    'Count': list(violent_top_2022.values) + [violent_other_2022] + list(violent_top_2023.values) + [violent_other_2023]
}

df_violent_data = pd.DataFrame(violent_data)

# Plot for Violent Crimes
fig_violent = px.bar(df_violent_data, x='Year', y='Count', color='Offense', barmode='group',
             title='Top Violent Crimes in 2022 vs 2023')
fig_violent.show()

# Repeat the same process for Non-Violent Crimes
non_violent_counts_2022 = df_2022[df_2022['Crime_Category'] == 'Non-Violent']['offense'].value_counts()
non_violent_counts_2023 = df_2023[df_2023['Crime_Category'] == 'Non-Violent']['offense'].value_counts()

non_violent_top_2022 = non_violent_counts_2022.nlargest(4)
non_violent_top_2023 = non_violent_counts_2023.nlargest(4)

non_violent_other_2022 = non_violent_counts_2022.sum() - non_violent_top_2022.sum()
non_violent_other_2023 = non_violent_counts_2023.sum() - non_violent_top_2023.sum()

# Prepare Data for Non-Violent Crimes plot
non_violent_data = {
    'Year': ['2022', '2022', '2022', '2022', '2022', '2023', '2023', '2023', '2023', '2023'],
    'Offense': list(non_violent_top_2022.index) + ['Other'] + list(non_violent_top_2023.index) + ['Other'],
    'Count': list(non_violent_top_2022.values) + [non_violent_other_2022] + list(non_violent_top_2023.values) + [non_violent_other_2023]
}

df_non_violent_data = pd.DataFrame(non_violent_data)

# Plot for Non-Violent Crimes
fig_non_violent = px.bar(df_non_violent_data, x='Year', y='Count', color='Offense', barmode='group',
             title='Top Non-Violent Crimes in 2022 vs 2023')
fig_non_violent.show()


In [40]:
import pandas as pd

# Compare violent offenses between 2022 and 2023
violent_counts_2022 = df_2022[df_2022['Crime_Category'] == 'Violent']['offense'].value_counts()
violent_counts_2023 = df_2023[df_2023['Crime_Category'] == 'Violent']['offense'].value_counts()

# Merge 2022 and 2023 data into a single dataframe for side-by-side comparison
violent_comparison = pd.DataFrame({
    '2022': violent_counts_2022,
    '2023': violent_counts_2023
}).fillna(0)  # Fill NaN with 0 for offenses not recorded in one year

# Calculate the percentage change (up or down)
violent_comparison['Change (%)'] = ((violent_comparison['2023'] - violent_comparison['2022']) / violent_comparison['2022']) * 100

# Avoid using `inplace=True`. Instead, assign the fixed column back.
violent_comparison['Change (%)'] = violent_comparison['Change (%)'].replace([float('inf'), -float('inf')], 0)
violent_comparison['Change (%)'] = violent_comparison['Change (%)'].fillna(0)  # Handle NaN if 2022 = 0

# Show the comparison with percentage change for violent crimes
print(violent_comparison)

# Repeat the process for non-violent offenses
non_violent_counts_2022 = df_2022[df_2022['Crime_Category'] == 'Non-Violent']['offense'].value_counts()
non_violent_counts_2023 = df_2023[df_2023['Crime_Category'] == 'Non-Violent']['offense'].value_counts()

non_violent_comparison = pd.DataFrame({
    '2022': non_violent_counts_2022,
    '2023': non_violent_counts_2023
}).fillna(0)  # Fill NaN with 0 for offenses not recorded in one year

# Calculate the percentage change (up or down)
non_violent_comparison['Change (%)'] = ((non_violent_comparison['2023'] - non_violent_comparison['2022']) / non_violent_comparison['2022']) * 100

# Avoid using `inplace=True`. Instead, assign the fixed column back.
non_violent_comparison['Change (%)'] = non_violent_comparison['Change (%)'].replace([float('inf'), -float('inf')], 0)
non_violent_comparison['Change (%)'] = non_violent_comparison['Change (%)'].fillna(0)  # Handle NaN if 2022 = 0

# Show the comparison with percentage change for non-violent crimes
print(non_violent_comparison)


         2022  2023  Change (%)
offense                        
ABSVAR      2     1  -50.000000
ADLTTN     16    20   25.000000
ASLT1      22    16  -27.272727
ASLT2    1221  1195   -2.129402
ASLT3     183   152  -16.939891
ASLT4      82   108   31.707317
CARJCK    518   300  -42.084942
CSCR      321   342    6.542056
DASLT1      1     2  100.000000
DASLT2    443   415   -6.320542
DASLT3     60    88   46.666667
DASTR     403   404    0.248139
JHOMIC      3     1  -66.666667
MURDR      77    69  -10.389610
ROBBIZ    115   100  -13.043478
ROBPAG    700   637   -9.000000
ROBPER    445   395  -11.235955
         2022  2023  Change (%)
offense                        
ARSON      98   126   28.571429
AUTOTH   6089  7694   26.359008
BIKETF   1125  1020   -9.333333
BURGB     943   800  -15.164369
BURGD    1635  1809   10.642202
COINOP      1     2  100.000000
COMPUT      5     3  -40.000000
DISARM      1     1    0.000000
MVTHFT     16    19   18.750000
NOPAY       1     1    0.000000
ONLTHT  

In [43]:
# Import necessary libraries
import pandas as pd

# Create a dictionary mapping each offense code to its description
offense_descriptions = {
    'ABSVAR': 'Sexual assault involving a vulnerable adult',
    'ADLTTN': 'Adult in need of protection/neglect',
    'ASLT1': 'Assault in the first degree',
    'ASLT2': 'Assault in the second degree',
    'ASLT3': 'Assault in the third degree',
    'ASLT4': 'Assault in the fourth degree',
    'CARJCK': 'Carjacking',
    'CSCR': 'Criminal sexual conduct/rape',
    'DASLT1': 'Domestic assault in the first degree',
    'DASLT2': 'Domestic assault in the second degree',
    'DASLT3': 'Domestic assault in the third degree',
    'DASTR': 'Domestic assault involving strangulation',
    'JHOMIC': 'Juvenile homicide',
    'MURDR': 'Murder',
    'ROBBIZ': 'Robbery of a business',
    'ROBPAG': 'Robbery with a gun',
    'ROBPER': 'Personal robbery',
    'ARSON': 'Arson',
    'AUTOTH': 'Auto theft',
    'BIKETF': 'Bicycle theft',
    'BURGB': 'Burglary of a business',
    'BURGD': 'Burglary of a dwelling',
    'COINOP': 'Coin-operated machine theft',
    'COMPUT': 'Computer-related crime',
    'DISARM': 'Disarming a police officer',
    'MVTHFT': 'Motor vehicle theft',
    'NOPAY': 'Failure to pay (theft)',
    'ONLTHT': 'Online theft',
    'PETIT': 'Petty theft',
    'POCKET': 'Pickpocketing',
    'SCRAP': 'Scrap metal theft',
    'SHOPLF': 'Shoplifting',
    'TBLDG': 'Theft from building',
    'TFMV': 'Theft from motor vehicle',
    'TFPER': 'Theft from person',
    'THEFT': 'General theft',
    'THFTSW': 'Theft of services',
    'TMVP': 'Theft of motor vehicle parts',
    'WLKOFF': 'Walking off without paying'
}

# Add the 'Description' column to both comparison DataFrames
violent_comparison['Description'] = violent_comparison.index.map(offense_descriptions)
non_violent_comparison['Description'] = non_violent_comparison.index.map(offense_descriptions)

# Determine the maximum length of any description for alignment
max_len = max(violent_comparison['Description'].map(len).max(), non_violent_comparison['Description'].map(len).max())

# Right-justify the 'Description' column to the maximum length
violent_comparison['Description'] = violent_comparison['Description'].str.rjust(max_len)
non_violent_comparison['Description'] = non_violent_comparison['Description'].str.rjust(max_len)

# Sort both DataFrames by the '2022' column in descending order
violent_comparison = violent_comparison.sort_values(by='2022', ascending=False)
non_violent_comparison = non_violent_comparison.sort_values(by='2022', ascending=False)

# Display the updated comparison DataFrames
print(violent_comparison)
print('\n')
print(non_violent_comparison)


         2022  2023  Change (%)                                  Description
offense                                                                     
ASLT2    1221  1195   -2.129402                 Assault in the second degree
ROBPAG    700   637   -9.000000                           Robbery with a gun
CARJCK    518   300  -42.084942                                   Carjacking
ROBPER    445   395  -11.235955                             Personal robbery
DASLT2    443   415   -6.320542        Domestic assault in the second degree
DASTR     403   404    0.248139     Domestic assault involving strangulation
CSCR      321   342    6.542056                 Criminal sexual conduct/rape
ASLT3     183   152  -16.939891                  Assault in the third degree
ROBBIZ    115   100  -13.043478                        Robbery of a business
ASLT4      82   108   31.707317                 Assault in the fourth degree
MURDR      77    69  -10.389610                                       Murder

In [48]:
import pandas as pd

# Create a dictionary mapping each offense code to its description
offense_descriptions = {
    'ABSVAR': 'Sexual assault involving a vulnerable adult',
    'ADLTTN': 'Adult in need of protection/neglect',
    'ASLT1': 'Assault in the first degree',
    'ASLT2': 'Assault in the second degree',
    'ASLT3': 'Assault in the third degree',
    'ASLT4': 'Assault in the fourth degree',
    'CARJCK': 'Carjacking',
    'CSCR': 'Criminal sexual conduct/rape',
    'DASLT1': 'Domestic assault in the first degree',
    'DASLT2': 'Domestic assault in the second degree',
    'DASLT3': 'Domestic assault in the third degree',
    'DASTR': 'Domestic assault involving strangulation',
    'JHOMIC': 'Juvenile homicide',
    'MURDR': 'Murder',
    'ROBBIZ': 'Robbery of a business',
    'ROBPAG': 'Robbery with a gun',
    'ROBPER': 'Personal robbery',
    'ARSON': 'Arson',
    'AUTOTH': 'Auto theft',
    'BIKETF': 'Bicycle theft',
    'BURGB': 'Burglary of a business',
    'BURGD': 'Burglary of a dwelling',
    'COINOP': 'Coin-operated machine theft',
    'COMPUT': 'Computer-related crime',
    'DISARM': 'Disarming a police officer',
    'MVTHFT': 'Motor vehicle theft',
    'NOPAY': 'Failure to pay (theft)',
    'ONLTHT': 'Online theft',
    'PETIT': 'Petty theft',
    'POCKET': 'Pickpocketing',
    'SCRAP': 'Scrap metal theft',
    'SHOPLF': 'Shoplifting',
    'TBLDG': 'Theft from building',
    'TFMV': 'Theft from motor vehicle',
    'TFPER': 'Theft from person',
    'THEFT': 'General theft',
    'THFTSW': 'Theft of services',
    'TMVP': 'Theft of motor vehicle parts',
    'WLKOFF': 'Walking off without paying'
}

# Add the 'Description' column to both comparison DataFrames
violent_comparison['Description'] = violent_comparison.index.map(offense_descriptions)
non_violent_comparison['Description'] = non_violent_comparison.index.map(offense_descriptions)

# Fill NaN values with empty strings to avoid 'float' errors
violent_comparison['Description'] = violent_comparison['Description'].fillna('')
non_violent_comparison['Description'] = non_violent_comparison['Description'].fillna('')

# Determine the maximum length of any description for alignment
max_len = max(violent_comparison['Description'].map(len).max(), non_violent_comparison['Description'].map(len).max())

# Right-justify the 'Description' column to the maximum length
violent_comparison['Description'] = violent_comparison['Description'].str.rjust(max_len)
non_violent_comparison['Description'] = non_violent_comparison['Description'].str.rjust(max_len)

# Sort both DataFrames by the '2022' column in descending order
violent_comparison = violent_comparison.sort_values(by='2022', ascending=False)
non_violent_comparison = non_violent_comparison.sort_values(by='2022', ascending=False)

# Calculate overall violent crime for 2022 and 2023
total_violent_2022 = violent_comparison['2022'].sum()
total_violent_2023 = violent_comparison['2023'].sum()
violent_change = ((total_violent_2023 - total_violent_2022) / total_violent_2022) * 100

# Calculate overall non-violent crime for 2022 and 2023
total_non_violent_2022 = non_violent_comparison['2022'].sum()
total_non_violent_2023 = non_violent_comparison['2023'].sum()
non_violent_change = ((total_non_violent_2023 - total_non_violent_2022) / total_non_violent_2022) * 100

# Create summary row for violent and non-violent crimes
summary_violent = pd.DataFrame({
    '2022': [total_violent_2022],
    '2023': [total_violent_2023],
    'Change (%)': [violent_change],
    'Description': ['Overall Violent Crime'.rjust(max_len)]
}, index=['Total'])

summary_non_violent = pd.DataFrame({
    '2022': [total_non_violent_2022],
    '2023': [total_non_violent_2023],
    'Change (%)': [non_violent_change],
    'Description': ['Overall Non-Violent Crime'.rjust(max_len)]
}, index=['Total'])

# Append the summary rows to the respective comparison DataFrames
violent_comparison = pd.concat([violent_comparison, summary_violent])
non_violent_comparison = pd.concat([non_violent_comparison, summary_non_violent])

# Display the updated comparison DataFrames
print(violent_comparison)
print('\n')
print(non_violent_comparison)


         2022   2023  Change (%)                                  Description
Total    9224   8490   -7.957502                                             
Total    4612   4245   -7.957502                                             
ASLT2    1221   1195   -2.129402                 Assault in the second degree
ROBPAG    700    637   -9.000000                           Robbery with a gun
CARJCK    518    300  -42.084942                                   Carjacking
ROBPER    445    395  -11.235955                             Personal robbery
DASLT2    443    415   -6.320542        Domestic assault in the second degree
DASTR     403    404    0.248139     Domestic assault involving strangulation
CSCR      321    342    6.542056                 Criminal sexual conduct/rape
ASLT3     183    152  -16.939891                  Assault in the third degree
ROBBIZ    115    100  -13.043478                        Robbery of a business
ASLT4      82    108   31.707317                 Assault in the 

In [38]:
import pandas as pd
import plotly.express as px

# Define the top 4 subcategories based on 2022 data for violent crimes
top_4_violent_offenses = ['ASLT2', 'ROBPAG', 'CARJCK', 'ROBPER']

# Create 'Other' category for both years for violent crimes
violent_counts_2022 = df_2022[df_2022['Crime_Category'] == 'Violent']['offense'].value_counts()
violent_counts_2023 = df_2023[df_2023['Crime_Category'] == 'Violent']['offense'].value_counts()

# Filter for the same top 4 subcategories for both years and sum the rest as "Other"
violent_top_2022 = violent_counts_2022.loc[top_4_violent_offenses]
violent_top_2023 = violent_counts_2023.loc[top_4_violent_offenses]

violent_other_2022 = violent_counts_2022.sum() - violent_top_2022.sum()
violent_other_2023 = violent_counts_2023.sum() - violent_top_2023.sum()

# Prepare data for Violent Crimes plot
violent_data = {
    'Year': ['2022', '2022', '2022', '2022', '2022', '2023', '2023', '2023', '2023', '2023'],
    'Offense': top_4_violent_offenses + ['Other'] + top_4_violent_offenses + ['Other'],
    'Count': list(violent_top_2022.values) + [violent_other_2022] + list(violent_top_2023.values) + [violent_other_2023]
}

df_violent_data = pd.DataFrame(violent_data)

# Plot for Violent Crimes
fig_violent = px.bar(df_violent_data, x='Year', y='Count', color='Offense', barmode='group',
             title='Top Violent Crimes in 2022 vs 2023 (Consistent Categories)')
fig_violent.show()

# Repeat the same process for Non-Violent Crimes
# Define the top 4 subcategories based on 2022 data for non-violent crimes
top_4_non_violent_offenses = ['AUTOTH', 'TFMV', 'THEFT', 'TMVP']

# Create 'Other' category for both years for non-violent crimes
non_violent_counts_2022 = df_2022[df_2022['Crime_Category'] == 'Non-Violent']['offense'].value_counts()
non_violent_counts_2023 = df_2023[df_2023['Crime_Category'] == 'Non-Violent']['offense'].value_counts()

# Filter for the same top 4 subcategories for both years and sum the rest as "Other"
non_violent_top_2022 = non_violent_counts_2022.loc[top_4_non_violent_offenses]
non_violent_top_2023 = non_violent_counts_2023.loc[top_4_non_violent_offenses]

non_violent_other_2022 = non_violent_counts_2022.sum() - non_violent_top_2022.sum()
non_violent_other_2023 = non_violent_counts_2023.sum() - non_violent_top_2023.sum()

# Prepare data for Non-Violent Crimes plot
non_violent_data = {
    'Year': ['2022', '2022', '2022', '2022', '2022', '2023', '2023', '2023', '2023', '2023'],
    'Offense': top_4_non_violent_offenses + ['Other'] + top_4_non_violent_offenses + ['Other'],
    'Count': list(non_violent_top_2022.values) + [non_violent_other_2022] + list(non_violent_top_2023.values) + [non_violent_other_2023]
}

df_non_violent_data = pd.DataFrame(non_violent_data)

# Plot for Non-Violent Crimes
fig_non_violent = px.bar(df_non_violent_data, x='Year', y='Count', color='Offense', barmode='group',
             title='Top Non-Violent Crimes in 2022 vs 2023 (Consistent Categories)')
fig_non_violent.show()


In [20]:
# Based
# Homicides comparison for 2022 vs 2023 (offense code: MURDR)
homicides_2022 = df_2022[df_2022['offense'] == 'MURDR']['offense'].count()
homicides_2023 = df_2023[df_2023['offense'] == 'MURDR']['offense'].count()

print(f"Homicides in 2022: {homicides_2022}")
print(f"Homicides in 2023: {homicides_2023}")

# Calculate percentage change
homicide_change = ((homicides_2023 - homicides_2022) / homicides_2022) * 100
print(f"Percentage change in homicides from 2022 to 2023: {homicide_change:.2f}%")


Homicides in 2022: 77
Homicides in 2023: 69
Percentage change in homicides from 2022 to 2023: -10.39%


# Filtering Data Based on the `reportedDate` Field for Dates Before April 23

This section explains how to filter crime incident data based on the `reportedDate` field. We will focus on incidents that were reported **before April 23** for both 2022 and 2023.

## Steps for Filtering

1. **Convert the `reportedDate` to a `datetime` object**:
   The `reportedDate` column may be stored as a string. We need to ensure it is in a `datetime` format so that we can properly filter based on dates.

2. **Filter rows before April 23 for both years**:
   After converting the column to a `datetime` object, we can filter the dataset to only include rows where the `reportedDate` is before April 23.

### Python Code

```python
import pandas as pd

# Ensure the 'reportedDate' field is in datetime format
df_2022['reportedDate'] = pd.to_datetime(df_2022['reportedDate'])
df_2023['reportedDate'] = pd.to_datetime(df_2023['reportedDate'])

# Define the cutoff date for filtering
cutoff_date = pd.to_datetime('2023-04-23')

# Filter data for both years to include only incidents before April 23
df_2022_filtered = df_2022[df_2022['reportedDate'] < cutoff_date.replace(year=2022)]
df_2023_filtered = df_2023[df_2023['reportedDate'] < cutoff_date.replace(year=2023)]

# Display the filtered data counts for both years
print(f"Filtered 2022 data before April 23: {len(df_2022_filtered)} rows")
print(f"Filtered 2023 data before April 23: {len(df_2023_filtered)} rows")


In [22]:
import pandas as pd

# Ensure that 'reportedDate' is in datetime format and convert to timezone-naive (remove timezone info)
df_2022['reportedDate'] = pd.to_datetime(df_2022['reportedDate']).dt.tz_localize(None)
df_2023['reportedDate'] = pd.to_datetime(df_2023['reportedDate']).dt.tz_localize(None)

# Define the cutoff date (timezone-naive, no need for timezone conversion)
cutoff_date = pd.to_datetime('2022-04-23')

# Filter for rows where reportedDate is before April 23 in both datasets
df_2022_filtered = df_2022[df_2022['reportedDate'] < cutoff_date.replace(year=2022)]
df_2023_filtered = df_2023[df_2023['reportedDate'] < cutoff_date.replace(year=2023)]

# Check filtered data counts
print(f"Filtered 2022 data before April 23: {len(df_2022_filtered)} rows")
print(f"Filtered 2023 data before April 23: {len(df_2023_filtered)} rows")


Filtered 2022 data before April 23: 6884 rows
Filtered 2023 data before April 23: 7604 rows


In [49]:
# Define violent and non-violent crimes based on the offense codes
violent_crimes = ['ASLT1', 'ASLT2', 'ASLT3', 'ASLT4', 'DASLT1', 'DASLT2', 'DASLT3', 'DASTR', 'CARJCK', 
                  'ROBPER', 'ROBBIZ', 'ROBPAG', 'MURDR', 'JHOMIC', 'CSCR', 'ABSVAR']
non_violent_crimes = ['THEFT', 'TFMV', 'TMVP', 'BIKETF', 'SHOPLF', 'THFTSW', 'ONLTHT', 'TBLDG', 'AUTOTH', 
                      'TFPER', 'MVTHFT', 'WLKOFF', 'BURGD', 'BURGB', 'ARSON', 'COMPUT', 'SCRAP', 'COINOP', 
                      'POCKET', 'NOPAY', 'DISARM', 'PETIT']

# Function to classify crimes into Violent or Non-Violent
def classify_crime(offense):
    if offense in violent_crimes:
        return 'Violent'
    elif offense in non_violent_crimes:
        return 'Non-Violent'
    else:
        return 'Other'

# Apply this classification to both filtered datasets
# Apply the classification to both filtered datasets using .loc to avoid the warning
df_2022_filtered.loc[:, 'Crime_Category'] = df_2022_filtered['offense'].apply(classify_crime)
df_2023_filtered.loc[:, 'Crime_Category'] = df_2023_filtered['offense'].apply(classify_crime)



Violent Crime Table:
   Offense                               Description   2022   2023 Change (%)
0    Total                                             6884   7604  10.459035
1    Total                                             1151   1040  -9.643788
2    ASLT2              Assault in the second degree    287    298   3.832753
3   ROBPAG                        Robbery with a gun    159    140 -11.949686
4   CARJCK                                Carjacking    152     91 -40.131579
5   ROBPER                          Personal robbery     94     92   -2.12766
6   DASLT2     Domestic assault in the second degree    121    107 -11.570248
7    DASTR  Domestic assault involving strangulation    113    125  10.619469
8     CSCR              Criminal sexual conduct/rape     86     89   3.488372
9    ASLT3               Assault in the third degree     52     36 -30.769231
10  ROBBIZ                     Robbery of a business     46     19 -58.695652
11   ASLT4              Assault in the four

In [27]:
import pandas as pd
import plotly.express as px

# Define the top 4 subcategories based on 2022 data for violent crimes
top_4_violent_offenses = ['ASLT2', 'ROBPAG', 'CARJCK', 'ROBPER']

# Create 'Other' category for both years for violent crimes
violent_counts_2022 = df_2022_filtered[df_2022_filtered['Crime_Category'] == 'Violent']['offense'].value_counts()
violent_counts_2023 = df_2023_filtered[df_2023_filtered['Crime_Category'] == 'Violent']['offense'].value_counts()

# Filter for the same top 4 subcategories for both years and sum the rest as "Other"
violent_top_2022 = violent_counts_2022.loc[top_4_violent_offenses]
violent_top_2023 = violent_counts_2023.loc[top_4_violent_offenses]

violent_other_2022 = violent_counts_2022.sum() - violent_top_2022.sum()
violent_other_2023 = violent_counts_2023.sum() - violent_top_2023.sum()

# Prepare data for Violent Crimes plot
violent_data = {
    'Year': ['2022', '2022', '2022', '2022', '2022', '2023', '2023', '2023', '2023', '2023'],
    'Offense': top_4_violent_offenses + ['Other'] + top_4_violent_offenses + ['Other'],
    'Count': list(violent_top_2022.values) + [violent_other_2022] + list(violent_top_2023.values) + [violent_other_2023]
}

df_violent_data = pd.DataFrame(violent_data)

# Plot for Violent Crimes
fig_violent = px.bar(df_violent_data, x='Year', y='Count', color='Offense', barmode='group',
             title='Top Violent Crimes in 2022 vs 2023 (Filtered, Consistent Categories)')
fig_violent.show()

# Repeat the same process for Non-Violent Crimes
# Define the top 4 subcategories based on 2022 data for non-violent crimes
top_4_non_violent_offenses = ['AUTOTH', 'TFMV', 'THEFT', 'TMVP']

# Create 'Other' category for both years for non-violent crimes
non_violent_counts_2022 = df_2022_filtered[df_2022_filtered['Crime_Category'] == 'Non-Violent']['offense'].value_counts()
non_violent_counts_2023 = df_2023_filtered[df_2023_filtered['Crime_Category'] == 'Non-Violent']['offense'].value_counts()

# Filter for the same top 4 subcategories for both years and sum the rest as "Other"
non_violent_top_2022 = non_violent_counts_2022.loc[top_4_non_violent_offenses]
non_violent_top_2023 = non_violent_counts_2023.loc[top_4_non_violent_offenses]

non_violent_other_2022 = non_violent_counts_2022.sum() - non_violent_top_2022.sum()
# Fix the variable name for non_violent_other_2023
non_violent_other_2023 = non_violent_counts_2023.sum() - non_violent_top_2023.sum()



In [50]:
import pandas as pd

# Offense descriptions (example, modify as needed)
offense_descriptions = {
    'ASLT2': 'Assault in the second degree',
    'ROBPAG': 'Robbery with a gun',
    'CARJCK': 'Carjacking',
    'ROBPER': 'Personal robbery',
    'AUTOTH': 'Auto theft',
    'TFMV': 'Theft from motor vehicle',
    'THEFT': 'General theft',
    'TMVP': 'Theft of motor vehicle parts',
    'Other': 'Other offenses'
}

# Create a function to generate the desired output for both violent and non-violent crimes
def create_crime_table(counts_2022, counts_2023, top_offenses, offense_type):
    # Filter for the same top 4 subcategories for both years and sum the rest as "Other"
    top_2022 = counts_2022.loc[top_offenses]
    top_2023 = counts_2023.loc[top_offenses]

    other_2022 = counts_2022.sum() - top_2022.sum()
    other_2023 = counts_2023.sum() - top_2023.sum()

    # Prepare data for output table
    data = {
        'Offense': top_offenses + ['Other'],
        'Description': [offense_descriptions.get(off, 'Unknown') for off in top_offenses] + ['Other offenses'],
        'Count 2022': list(top_2022.values) + [other_2022],
        'Count 2023': list(top_2023.values) + [other_2023]
    }

    # Calculate percentage change
    data['% Change'] = [(data['Count 2023'][i] - data['Count 2022'][i]) / data['Count 2022'][i] * 100
                        if data['Count 2022'][i] > 0 else 0 for i in range(len(data['Count 2022']))]

    # Convert to DataFrame
    df = pd.DataFrame(data)

    # Calculate grand totals
    total_2022 = df['Count 2022'].sum()
    total_2023 = df['Count 2023'].sum()
    total_change = (total_2023 - total_2022) / total_2022 * 100 if total_2022 > 0 else 0

    # Add grand totals as a row, separated by a blank line
    total_row = pd.DataFrame({
        'Offense': [''],
        'Description': ['Grand Total'],
        'Count 2022': [total_2022],
        'Count 2023': [total_2023],
        '% Change': [total_change]
    })

    df = pd.concat([df, pd.DataFrame([{'Offense': '', 'Description': '', 'Count 2022': '', 'Count 2023': '', '% Change': ''}]), total_row], ignore_index=True)

    return df

# Generate tables for violent and non-violent crimes
violent_crime_table = create_crime_table(violent_counts_2022, violent_counts_2023, top_4_violent_offenses, 'Violent')
non_violent_crime_table = create_crime_table(non_violent_counts_2022, non_violent_counts_2023, top_4_non_violent_offenses, 'Non-Violent')

# Display the tables
print("Violent Crime Table:")
print(violent_crime_table)

print("\nNon-Violent Crime Table:")
print(non_violent_crime_table)


Violent Crime Table:
  Offense                   Description Count 2022 Count 2023   % Change
0   ASLT2  Assault in the second degree       1221       1195  -2.129402
1  ROBPAG            Robbery with a gun        700        637       -9.0
2  CARJCK                    Carjacking        518        300 -42.084942
3  ROBPER              Personal robbery        445        395 -11.235955
4   Other                Other offenses       1728       1718  -0.578704
5                                                                       
6                           Grand Total       4612       4245  -7.957502

Non-Violent Crime Table:
  Offense                   Description Count 2022 Count 2023   % Change
0  AUTOTH                    Auto theft       6089       7694  26.359008
1    TFMV      Theft from motor vehicle       3613       3682    1.90977
2   THEFT                 General theft       3827       3552  -7.185785
3    TMVP  Theft of motor vehicle parts       3238       1806  -44.22483
4   

Violent crimes in 2022 (before April 23): 1182
Violent crimes in 2023 (before April 23): 1071
Non-violent crimes in 2022 (before April 23): 5699
Non-violent crimes in 2023 (before April 23): 6524
Percentage change in violent crimes from 2022 to 2023: -9.39%
Percentage change in non-violent crimes from 2022 to 2023: 14.48%


In [55]:
import pandas as pd

# Define violent and non-violent offenses
violent_offenses = ['ASLT1', 'ASLT2', 'ASLT3', 'ASLT4', 'CSCR', 'DASLT1', 'DASLT2', 'DASLT3', 'DASTR', 
                    'CARJCK', 'ROBBIZ', 'ROBPAG', 'ROBPER', 'MURDR', 'JHOMIC', 'ABSVAR']
non_violent_offenses = ['AUTOTH', 'BIKETF', 'BURGB', 'BURGD', 'SHOPLF', 'TFMV', 'TMVP', 'THEFT', 'THFTSW', 
                        'TFPER', 'TBLDG', 'ARSON', 'ONLTHT', 'MVTHFT', 'WLKOFF', 'COMPUT', 'PETIT', 'COINOP']

# Function to create crime table filtered by crime category (violent or non-violent) and sorted by 2022 counts
def create_crime_table(df_2022, df_2023, offenses, description):
    counts_2022 = df_2022[df_2022['offense'].isin(offenses)]['offense'].value_counts()
    counts_2023 = df_2023[df_2023['offense'].isin(offenses)]['offense'].value_counts()

    # Create a table for the given offenses
    offense_list = sorted(set(counts_2022.index).union(set(counts_2023.index)))

    data = {
        'Offense': offense_list,
        'Description': [offense_descriptions.get(off, 'Unknown') for off in offense_list],
        '2022': [counts_2022.get(off, 0) for off in offense_list],
        '2023': [counts_2023.get(off, 0) for off in offense_list]
    }

    # Calculate percentage change
    data['Change (%)'] = [(data['2023'][i] - data['2022'][i]) / data['2022'][i] * 100 if data['2022'][i] > 0 else 0 for i in range(len(data['2022']))]

    # Convert to DataFrame
    df_table = pd.DataFrame(data)

    # Sort by 2022 counts in descending order
    df_table = df_table.sort_values(by='2022', ascending=False).reset_index(drop=True)

    # Add overall total row at the bottom
    total_row = pd.DataFrame({
        'Offense': ['Total'],
        'Description': [description],
        '2022': [counts_2022.sum()],
        '2023': [counts_2023.sum()],
        'Change (%)': [(counts_2023.sum() - counts_2022.sum()) / counts_2022.sum() * 100 if counts_2022.sum() > 0 else 0]
    })

    # Concatenate with the total row
    df_table = pd.concat([df_table, total_row], ignore_index=True)
    
    return df_table

# Create tables for violent and non-violent crimes
violent_crime_table = create_crime_table(df_2022_filtered, df_2023_filtered, violent_offenses, 'Overall Violent Crime')
non_violent_crime_table = create_crime_table(df_2022_filtered, df_2023_filtered, non_violent_offenses, 'Overall Non-Violent Crime')

# Display the tables
print("Violent Crime Table:")
print(violent_crime_table)

print("\nNon-Violent Crime Table:")
print(non_violent_crime_table)


Violent Crime Table:
   Offense                               Description  2022  2023  Change (%)
0    ASLT2              Assault in the second degree   287   298    3.832753
1   ROBPAG                        Robbery with a gun   159   140  -11.949686
2   CARJCK                                Carjacking   152    91  -40.131579
3   DASLT2     Domestic assault in the second degree   121   107  -11.570248
4    DASTR  Domestic assault involving strangulation   113   125   10.619469
5   ROBPER                          Personal robbery    94    92   -2.127660
6     CSCR              Criminal sexual conduct/rape    86    89    3.488372
7    ASLT3               Assault in the third degree    52    36  -30.769231
8   ROBBIZ                     Robbery of a business    46    19  -58.695652
9    MURDR                                    Murder    22    15  -31.818182
10   ASLT4              Assault in the fourth degree    19    28   47.368421
11  DASLT3                                   Unknown   